### PR TITLE
LPS-151265-LPS-151267 key version auto generation and upgrade process

### PR DESCRIPTION
**Pull request by wcao20170619**
https://issues.liferay.com/browse/LPS-151265
https://issues.liferay.com/browse/LPS-151267

The PR combines the above 2 tickets.
It has taken all @peerkar feedback from the tickets and [PR](https://github.com/liferay-search/liferay-portal/pull/426)  

@xbrianlee will run the Database Upgrade Tool to verify the UpgradeProcess working ok with existing  SXPBlueprint/SXPElement data. - thank you!

### CODE REVIEW

**Comment by wcao20170619** [comment]
ci:test:relevant

**Comment by timpak** [comment]
The test failures from the ci:test:search run are known failures and aren't caused by the changes in the PR. Should be good to forward once it passes review.

**Comment by wcao20170619** [comment]
ci:test:sf

@peerkar, I also split Key/Version data upgrade processes of SXPBlueprint/SXPElement into two processes as you suggested in the ticket. And tested the upgrade process by using the Database Upgrade Tool and environment variable (LIFERAY_UPGRADE_PERIOD_DATABASE_PERIOD_AUTO_PERIOD_RUN=true) with @xbrianlee helps.

**Comment by wcao20170619** [comment]
ci:test:search

**Comment by wcao20170619** [comment]
ci:test:relevant

File: modules/dxp/apps/search-experiences/search-experiences-service/src/main/java/com/liferay/search/experiences/service/impl/SXPBlueprintLocalServiceImpl.java [hunk 1]
```diff
@@ -77,8 +78,13 @@ public SXPBlueprint addSXPBlueprint(
 		sxpBlueprint.setConfigurationJSON(configurationJSON);
 		sxpBlueprint.setDescriptionMap(descriptionMap);
 		sxpBlueprint.setElementInstancesJSON(elementInstancesJSON);
-		sxpBlueprint.setSchemaVersion(schemaVersion);
+		sxpBlueprint.setKey(KeyUtil.getKey(counterLocalService, null));
 		sxpBlueprint.setTitleMap(titleMap);
+		sxpBlueprint.setSchemaVersion(schemaVersion);
+		sxpBlueprint.setVersion(
```
Comment (BryanEngler) [inline modules/dxp/apps/search-experiences/search-experiences-service/src/main/java/com/liferay/search/experiences/service/impl/SXPBlueprintLocalServiceImpl.java:84]: could we just put setVersion("1.0") here?

File: modules/dxp/apps/search-experiences/search-experiences-test/src/testIntegration/java/com/liferay/search/experiences/service/impl/test/SXPBlueprintLocalServiceImplTest.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+			_sxpBlueprint.getUserId(), _sxpBlueprint.getSXPBlueprintId(),
+			_sxpBlueprint.getConfigurationJSON(),
+			_sxpBlueprint.getDescriptionMap(),
+			_sxpBlueprint.getElementInstancesJSON(),
+			_sxpBlueprint.getSchemaVersion(), _sxpBlueprint.getTitleMap(),
+			ServiceContextTestUtil.getServiceContext(
+				_group, TestPropsValues.getUserId()));
+
+		Assert.assertEquals("1.1", _sxpBlueprint.getVersion());
+		Assert.assertNotEquals(key, _sxpBlueprint.getKey());
```
Comment (BryanEngler) [inline modules/dxp/apps/search-experiences/search-experiences-test/src/testIntegration/java/com/liferay/search/experiences/service/impl/test/SXPBlueprintLocalServiceImplTest.java:88]: SF - add a line break?

File: modules/dxp/apps/search-experiences/search-experiences-test/src/testIntegration/java/com/liferay/search/experiences/service/impl/test/SXPBlueprintLocalServiceImplTest.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+			_sxpBlueprint.getUserId(), _sxpBlueprint.getSXPBlueprintId(),
+			_sxpBlueprint.getConfigurationJSON(),
+			_sxpBlueprint.getDescriptionMap(),
+			_sxpBlueprint.getElementInstancesJSON(),
+			_sxpBlueprint.getSchemaVersion(), _sxpBlueprint.getTitleMap(),
+			ServiceContextTestUtil.getServiceContext(
+				_group, TestPropsValues.getUserId()));
+
+		Assert.assertEquals("1.2", _sxpBlueprint.getVersion());
+
```
Comment (BryanEngler) [inline modules/dxp/apps/search-experiences/search-experiences-test/src/testIntegration/java/com/liferay/search/experiences/service/impl/test/SXPBlueprintLocalServiceImplTest.java:100]: SF - remove a line break?

**Comment by BryanEngler** [comment]
looks pretty good, the setVersion logic looked a little more complicated than i think was necessary... and the line breaks seemed inconsistent in the tests. if you fix those couple of things up i think we can forward this

File: modules/dxp/apps/search-experiences/search-experiences-test/src/testIntegration/java/com/liferay/search/experiences/service/impl/test/SXPBlueprintLocalServiceImplTest.java (hunk 2 above)
Comment (wcao20170619) [reply modules/dxp/apps/search-experiences/search-experiences-test/src/testIntegration/java/com/liferay/search/experiences/service/impl/test/SXPBlueprintLocalServiceImplTest.java:89]: added a line break - thank you!

File: modules/dxp/apps/search-experiences/search-experiences-test/src/testIntegration/java/com/liferay/search/experiences/service/impl/test/SXPBlueprintLocalServiceImplTest.java (hunk 3 above)
Comment (wcao20170619) [reply modules/dxp/apps/search-experiences/search-experiences-test/src/testIntegration/java/com/liferay/search/experiences/service/impl/test/SXPBlueprintLocalServiceImplTest.java:101]: kept a line break - thank you!

**Comment by wcao20170619** [comment]
ci:test:search

**Comment by wcao20170619** [comment]
ci:test:sf

**Comment by wcao20170619** [comment]
ci:test:relevant

**Comment by wcao20170619** [comment]
ci:test:search

**Comment by wcao20170619** [comment]
ci:test:relevant

**Comment by wcao20170619** [comment]
ci:forward

**Comment by wcao20170619** [comment]
ci:test:sf

**Comment by wcao20170619** [comment]
ci:forward

**Comment by wcao20170619** [comment]
ci:test:sf

**Comment by wcao20170619** [comment]
ci:test:relevant

**Comment by wcao20170619** [comment]
ci:test:relevant

**Comment by wcao20170619** [comment]
continue to https://github.com/liferay-search/liferay-portal/pull/439

**Pull request closed without merging**